src/modules/flow/string/string-ascii.c

```diff
@@ -146,27 +181,45 @@ string_concat(struct sol_flow_node *node,
 {
     struct string_concatenate_data *mdata = data;
     char *dest;
-    int r, len;
+    int r;
+    uint8_t i, count = 0;
+    size_t len = 0, sep_len = 0;
 
-    if (!get_string_by_port(packet, port, &mdata->base))
+    r = get_string_by_port(packet, port, mdata->string);
+    SOL_INT_CHECK(r, < 0, r);
+
+    mdata->var_initialized |= 1u << port;
+    if (mdata->var_initialized != mdata->var_connected)
         return 0;
 
-    len = strlen(mdata->base.string[0]) + strlen(mdata->base.string[1]) + 1;
+    //calculate final length
     if (mdata->separator)
-        len += strlen(mdata->separator);
+        sep_len = strlen(mdata->separator);
+    len = string_list_len(mdata->string, mdata->var_initialized, sep_len,
+        mdata->n);
 
-    dest = calloc(len, sizeof(*dest));
+    if (len > SIZE_MAX - 1)
+        return -EOVERFLOW;
+    dest = calloc(len + 1, sizeof(*dest));
     SOL_NULL_CHECK(dest, -ENOMEM);
 
-    dest = strcpy(dest, mdata->base.string[0]);
+    for (i = 0; i < 32; i++) {
+        if (!(mdata->var_initialized & (1u << i)))
+            continue;
 
-    if (mdata->separator)
-        dest = strcat(dest, mdata->separator);
+        if (count && mdata->separator) {
+            dest = strncat(dest, mdata->separator, len);
```

Comment:
I was planning on using buffer, but then I just followed the approach that was being used before. But I think that it really better use buffer for that.